server/backend/api/app/crud/services/ext_app_user_search_service.py

```diff
@@ -0,0 +1,218 @@
+                                            ExtUserSearchParamSchema)
+from api.app.schemas.ext.user_search import (ExtApplicationUserSearchGetSchema,
+                                             ExtApplicationUserSearchSchema,
+                                             ExtRoleWithScopeSchema)
+from api.app.schemas.requester import RequesterSchema
+from api.app.utils.utils import raise_http_exception
+from sqlalchemy import and_, func, or_, select
+from sqlalchemy.orm import Session, joinedload
+from sqlalchemy.sql.selectable import Select
+
+from .ext_api_interface import ExtAPIInterface
+
+LOGGER = logging.getLogger(__name__)
+
+class ExtAppUserSearchService(ExtAPIInterface):
+    """
+    Service to handle external application user search requests for external API calls.
+    The service only allows requesters with proper application role (with call_api permission)
+    to perform the search.
+    Ref API spec at https://apps.nrs.gov.bc.ca/int/confluence/display/FSAST1/Users+Search+API
+    """
+
+    def __init__(self, db: Session, requester: RequesterSchema, application_id: int, *args, **kwargs):
+        super().__init__(requester, application_id, db=db, *args, **kwargs)
+
+    def search_users(
+        self,
+        page_params: ExtUserSearchParamSchema,
+        filter_params: ExtApplicationUserSearchSchema
+    ) -> ExtUserSearchPagedResultsSchema:
+        """
+        For external Consumer API only.
+        Searches users associated with the application based on filter_params and paginates results based on page_params.
+        """
+        if not self.is_request_allowed():
+            error_msg = ("Programming error. This method is for external API only. "
+                         "Router should have already checked the requester call API permission.")
+            raise_http_exception(
+                status_code=HTTPStatus.INTERNAL_SERVER_ERROR,
+                error_code=ERROR_CODE_INVALID_OPERATION,
+                error_msg=error_msg,
+            )
+
+        LOGGER.debug(f"Searching users with filter_params: {filter_params}, page_params: {page_params}")
+
+        # Base select statement
+        user_role_stmt: Select = (
+            select(FamUser)
+            .join(FamUserRoleXref, FamUser.user_id == FamUserRoleXref.user_id)
+            .join(FamRole, FamUserRoleXref.role_id == FamRole.role_id)
+            .where(FamRole.application_id == self.application_id)
+        )
+
+        user_role_stmt: Select = self._apply_user_filters(user_role_stmt, filter_params)
+
+        # Get total count of distinct users
+        user_id_stmt: Select = user_role_stmt.with_only_columns(FamUser.user_id).distinct()
+        total = self.db.execute(select(func.count()).select_from(user_id_stmt.subquery())).scalar()
+        page = page_params.page or EXT_MIN_PAGE
+        size = page_params.size or EXT_MIN_PAGE_SIZE
+        page_count: int = (total + size - 1) // size
```

Comment:
This calculation looks wrong. max # of pages = total count / page size, rounded up.

Reply:
Hi Basil, this was an interesting one as it was AI provided trick and I evaluated its correctness when leaving it there.
For example:
Formula 1: pageCount = (total + size - 1) // size
Formula 2: pageCount = math.ceil(total / size)

* total = 0, size = 10
	page(Formula 1): 0
	page(Formula 2): 0

* total = 20, size = 10
	page(Formula 1): 2
	page(Formula 2): 2
	
* total = 45, size = 10
	page(Formula 1): 5
	page(Formula 2): 5

But I will change it since it might be better for readability.

Reply:
Ah, I did misread the original code, in part due to my limited familiarity with Python.
Yes, looks like the AI version is avoiding floating point math which is reasonable, but less clear for sure.